tests/Progress/ProgressUpdateRepositoryTest.php

```diff
@@ -0,0 +1,183 @@
+                            $progress->setPercentage(50);
+                            return $progress;
+                        })
+                    ];
+                }
+            ],
+            'Return two set of Activities with different complete statuses' => [
+                function () {
+                    return [
+                        $this->createProgress(),
+                        $this->createProgress(function (Progress $progress): Progress {
+                            $progress->setActivityId(factory(Activity::class)->create()->id);
+                            $progress->isComplete(false);
+                            return $progress;
+                        })
+                    ];
+                }
+            ]
+        ];
+    }
+
+    public function createProgress(\Closure $callback = null): Progress
+    {
+        $activity = factory(Activity::class)->create();
+        $modules = factory(ModuleInstance::class, 5)->create(['activity_id' => $activity->id]);
+
+        $progress = Progress::create(
+            $activity->id,
+            factory(ActivityInstance::class)->create()->id,
+            Carbon::now(),
+            true,
+            80
+        );
+
+        $moduleProgress1 = ModuleInstanceProgress::create(
+            $modules[0]->id,
+            true, true, 100, true, true
+        );
+        $moduleProgress2 = ModuleInstanceProgress::create(
+            $modules[0]->id,
+            true, true, 100, true, true
+        );
+        $progress->pushModule($moduleProgress1);
+        $progress->pushModule($moduleProgress2);
+
+        if($callback === null) {
+            $callback = fn(Progress $progress): Progress => $progress;
+        }
+
+        return $callback($progress);
+    }
+
+    /** @test */
+    public function hasChanged_returns_false_if_progress_has_not_changed_since_last_save()
+    {
+        $progress = $this->createProgress();
+        // Ensure that first pass returns true:
+        $this->assertTrue($this->repository->hasChanged($this->activityInstance->id, 'caller', $progress));
+        // Trigger Save of Progress:
+        $this->repository->saveChanges($this->activityInstance->id, 'caller', $progress);
+        // Ensure that second pass is false as no changes
+        $this->assertFalse($this->repository->hasChanged($this->activityInstance->id, 'caller', $progress));
+    }
+
+    /**
+     *
+     * @test
+     */
+    public function hasChanged_returns_true_if_progress_has_changed_since_last_save()
+    {
+        $progress = $this->createProgress();
+
+        // Ensure that first pass returns true:
+        $this->assertTrue($this->repository->hasChanged($this->activityInstance->id, 'caller', $progress));
+        // Trigger Save of Progress:
+        $this->repository->saveChanges($this->activityInstance->id, 'caller', $progress);
+
+        $updatedProgress = $this->createProgress(function (Progress $progress): Progress {
```

Comment:
Rather than creating a new progress, we want to test if the current progress changing affects the repository. You can just do something like `$progress->setComplete(false);` which will simulate a progress change much better